File: python/squirrelparser/ast_node.py
"""AST node representation for the Squirrel Parser."""

from __future__ import annotations
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .match_result import MatchResult

from .terminals import Str, Char, CharRange, AnyChar
from .combinators import Ref
from .match_result import SyntaxError as SyntaxErrorNode


@dataclass
class ASTNode:
    """An AST node representing either a rule match or a terminal match.

    AST nodes have:
    - A label (the rule name or terminal type)
    - A position and length in the input
    - Children (other AST nodes)
    - The matched text
    """

    label: str
    pos: int
    len: int
    children: list[ASTNode]
    _input: str

    @property
    def text(self) -> str:
        """Get the text matched by this node."""
        return self._input[self.pos:self.pos + self.len]

    def get_child(self, index: int) -> ASTNode:
        """Get a child by index."""
        return self.children[index]

    def __str__(self) -> str:
        return f'ASTNode({self.label}, "{self.text}", children: {len(self.children)})'

    def to_pretty_string(self, indent: int = 0) -> str:
        """Pretty print the AST tree."""
        lines = []
        prefix = '  ' * indent
        if self.children:
            lines.append(f'{prefix}{self.label}')
        else:
            lines.append(f'{prefix}{self.label}: "{self.text}"')

        for child in self.children:
            lines.append(child.to_pretty_string(indent + 1))

        return '\n'.join(lines)


def build_ast(match: MatchResult | None, input_str: str, top_rule: str | None = None) -> ASTNode | None:
    """Build an AST from a parse tree.

    The AST only includes:
    - Ref nodes (rule references) with their rule name as label
    - Terminal nodes (Str, Char, CharRange, AnyChar) with text as label

    All intermediate combinator nodes (Seq, First, etc.) are flattened out,
    with their children promoted to be children of the nearest ancestral rule.

    For top-level combinator matches, creates a synthetic node with the given top_rule label.

    Args:
        match: The match result
        input_str: The input string
        top_rule: Optional rule name for synthetic top-level nodes (combinators)

    Returns:
        The AST node, or None if match is None/mismatch and no top_rule provided
    """
    if match is None or match.is_mismatch:
        return None

    ast = _build_ast_node(match, input_str)

    # If top-level match is a combinator and we have a top_rule label, build synthetic node
    if ast is None and top_rule:
        children = _collect_children_for_ast(match, input_str)
        ast = ASTNode(
            label=top_rule,
            pos=match.pos,
            len=match.len,
            children=children,
            _input=input_str,
        )

    return ast


def _build_ast_node(match: MatchResult, input_str: str) -> ASTNode | None:
    """Build an AST node from a match result."""
    clause = match.clause

    # Handle Ref nodes - these become AST nodes with the rule name as label
    # UNLESS they're marked as transparent, in which case we flatten them
    if isinstance(clause, Ref):
        if clause.transparent:
            # Transparent rule - don't create a node, just return None
            return None
        # Get children by recursively processing the wrapped match
        children = _collect_children(match, input_str)
        return ASTNode(
            label=clause.rule_name,
            pos=match.pos,
            len=match.len,
            children=children,
            _input=input_str,
        )

    # Handle terminal nodes - these become leaf AST nodes
    if isinstance(clause, (Str, Char, CharRange, AnyChar)):
        return ASTNode(
            label=type(clause).__name__,
            pos=match.pos,
            len=match.len,
            children=[],
            _input=input_str,
        )

    # For all other nodes (combinators), flatten and collect children
    # This shouldn't normally be called at the top level, but handle it anyway
    return None


def _collect_children(match: MatchResult, input_str: str) -> list[ASTNode]:
    """Collect children for an AST node by flattening combinators."""
    result = []

    for child in match.sub_clause_matches:
        if child.is_mismatch:
            continue
        if isinstance(child, SyntaxErrorNode):  # Skip error nodes in AST
            continue

        clause = child.clause

        # If child is a Ref or terminal, add it as an AST node
        if isinstance(clause, (Ref, Str, Char, CharRange, AnyChar)):
            node = _build_ast_node(child, input_str)
            if node is not None:
                result.append(node)
        # Otherwise, it's a combinator - recursively collect its children
        else:
            result.extend(_collect_children(child, input_str))

    return result


def _collect_children_for_ast(match: MatchResult, input_str: str) -> list[ASTNode]:
    """Collect children for AST building when top-level match is a combinator.

    Used by parse_to_ast() to build synthetic AST nodes.
    """
    result = []

    for child in match.sub_clause_matches:
        if child.is_mismatch:
            continue
        if isinstance(child, SyntaxErrorNode):  # Skip error nodes in AST
            continue

        clause = child.clause

        # If child is a Ref, create an AST node for it (unless transparent)
        if isinstance(clause, Ref):
            if not clause.transparent:
                children = _collect_children(child, input_str)
                node = ASTNode(
                    label=clause.rule_name,
                    pos=child.pos,
                    len=child.len,
                    children=children,
                    _input=input_str,
                )
                result.append(node)
            # Transparent rules are completely skipped - don't create node and don't include their children
        # Include terminals as leaf nodes
        elif isinstance(clause, (Str, Char, CharRange, AnyChar)):
            node = _build_ast_node(child, input_str)
            if node is not None:
                result.append(node)
        # For other combinators, recursively collect their children
        else:
            result.extend(_collect_children_for_ast(child, input_str))

    return result

File: dart/lib/src/parser/ast_node.dart
import 'combinators.dart';
import 'match_result.dart';
import 'terminals.dart';

/// An AST node representing either a rule match or a terminal match.
///
/// AST nodes have:
/// - A label (the rule name or terminal type)
/// - A position and length in the input
/// - Children (other AST nodes)
/// - The matched text
class ASTNode {
  /// The label for this node (rule name or terminal type).
  final String label;

  /// The position in the input where this match starts.
  final int pos;

  /// The length of the match.
  final int len;

  /// Child AST nodes. These are flattened - only rule matches and terminals appear.
  final List<ASTNode> children;

  /// The input string (shared reference).
  final String _input;

  ASTNode({
    required this.label,
    required this.pos,
    required this.len,
    required this.children,
    required String input,
  }) : _input = input;

  /// Get the text matched by this node.
  String get text => _input.substring(pos, pos + len);

  /// Get a child by index.
  ASTNode getChild(int index) => children[index];

  @override
  String toString() => 'ASTNode($label, "$text", children: ${children.length})';

  /// Pretty print the AST tree.
  String toPrettyString({int indent = 0}) {
    final buffer = StringBuffer();
    buffer.write('  ' * indent);
    buffer.write(label);
    if (children.isEmpty) {
      buffer.write(': "$text"');
    }
    buffer.writeln();
    for (final child in children) {
      buffer.write(child.toPrettyString(indent: indent + 1));
    }
    return buffer.toString();
  }
}

/// Build an AST from a parse tree.
///
/// The AST only includes:
/// - Ref nodes (rule references) with their rule name as label
/// - Terminal nodes (Str, Char, CharRange, AnyChar) with text as label
///
/// All intermediate combinator nodes (Seq, First, etc.) are flattened out,
/// with their children promoted to be children of the nearest ancestral rule.
///
/// For top-level combinator matches, creates a synthetic node with the given topRule label.
ASTNode? buildAST(MatchResult? match, String input, {String? topRule}) {
  if (match == null || match.isMismatch) return null;

  var ast = _buildASTNode(match, input);

  // If top-level match is a combinator and we have a topRule label, build synthetic node
  if (ast == null && topRule != null) {
    final children = collectChildrenForAST(match, input);
    ast = ASTNode(
      label: topRule,
      pos: match.pos,
      len: match.len,
      children: children,
      input: input,
    );
  }

  return ast;
}

ASTNode? _buildASTNode(MatchResult match, String input) {
  final clause = match.clause;

  // Handle Ref nodes - these become AST nodes with the rule name as label
  // UNLESS they're marked as transparent, in which case we flatten them
  if (clause is Ref) {
    if (clause.transparent) {
      // Transparent rule - don't create a node, just return children
      return null;
    }
    // Get children by recursively processing the wrapped match
    final children = _collectChildren(match, input);
    return ASTNode(
      label: clause.ruleName,
      pos: match.pos,
      len: match.len,
      children: children,
      input: input,
    );
  }

  // Handle terminal nodes - these become leaf AST nodes
  if (clause is Terminal) {
    return ASTNode(
      label: clause.runtimeType.toString(),
      pos: match.pos,
      len: match.len,
      children: [],
      input: input,
    );
  }

  // For all other nodes (combinators), flatten and collect children
  // This shouldn't normally be called at the top level, but handle it anyway
  return null;
}

/// Collect children for an AST node by flattening combinators.
List<ASTNode> _collectChildren(MatchResult match, String input) {
  final result = <ASTNode>[];

  for (final child in match.subClauseMatches) {
    if (child.isMismatch || child is SyntaxError) {
      continue; // Skip error nodes in AST
    }

    final clause = child.clause;

    // If child is a Ref, terminal, or syntax error, add it as an AST node
    if (clause is Ref || clause is Terminal) {
      final node = _buildASTNode(child, input);
      if (node != null) {
        result.add(node);
      }
    }
    // Otherwise, it's a combinator - recursively collect its children
    else {
      result.addAll(_collectChildren(child, input));
    }
  }

  return result;
}

/// Collect children for AST building when top-level match is a combinator.
/// This is used by squirrelParse to build synthetic AST nodes from top-level combinators.
List<ASTNode> collectChildrenForAST(MatchResult match, String input) {
  final result = <ASTNode>[];

  for (final child in match.subClauseMatches) {
    if (child.isMismatch || child is SyntaxError) {
      continue; // Skip error nodes in AST
    }

    final clause = child.clause;

    // If child is a Ref, create an AST node for it (unless transparent)
    if (clause is Ref) {
      if (!clause.transparent) {
        final children = _collectChildren(child, input);
        result.add(ASTNode(
          label: clause.ruleName,
          pos: child.pos,
          len: child.len,
          children: children,
          input: input,
        ));
      }
      // Transparent rules are completely skipped - don't create node and don't include their children
    }
    // Include terminals as leaf nodes
    else if (clause is Terminal) {
      final node = _buildASTNode(child, input);
      if (node != null) {
        result.add(node);
      }
    }
    // For other combinators, recursively collect their children
    else {
      result.addAll(collectChildrenForAST(child, input));
    }
  }

  return result;
}
